third_party/move/move-compiler-v2/src/env_pipeline/model_ast_lints/blocks_in_conditions.rs

```diff
@@ -0,0 +1,98 @@
+//!
+//! We also only report on the outermost condition with blocks.
+
+use crate::env_pipeline::model_ast_lints::ExpressionLinter;
+use codespan_reporting::diagnostic::Severity;
+use move_model::{
+    ast::ExpData,
+    model::{GlobalEnv, NodeId},
+};
+
+/// Expression linter keeping track of traversal state.
+#[derive(Default)]
+pub struct BlocksInConditions {
+    /// Is `None` if we are not traversing a condition.
+    state: Option<CondExprState>,
+}
+
+/// State of the traversal when a condition expression has been found.
+enum CondExprState {
+    /// Condition expression with `id` should be examined for:
+    ///   the presence of blocks/sequences and the presence of spec blocks.
+    Examine { id: NodeId },
+    /// Traversing within a condition with `id`. During the traversal:
+    ///   `has_spec_block` is true if we have seen a spec block so far.
+    ///   `has_any_block` is true if we have seen any block/sequence so far.
+    Traversing {
+        id: NodeId,
+        has_any_block: bool,
+        has_spec_block: bool,
+    },
+}
+
+impl ExpressionLinter for BlocksInConditions {
+    fn visit_expr_pre(&mut self, expr: &ExpData, _env: &GlobalEnv) {
+        use CondExprState::*;
+        use ExpData::{Block, IfElse, Match, Sequence, SpecBlock};
+        match self.state {
+            None => {
+                if let IfElse(_, cond, _, _) | Match(_, cond, _) = expr {
+                    self.state = Some(Examine { id: cond.node_id() });
+                }
+            },
+            Some(Examine { id }) if expr.node_id() == id => {
+                // We are now starting to traverse a condition.
+                let has_any_block = matches!(expr, Block(..) | Sequence(..));
+                self.state = Some(Traversing {
+                    id,
+                    has_any_block,
+                    has_spec_block: false, // A spec block cannot appear directly in a condition.
+                });
+            },
+            Some(Traversing {
+                id,
+                has_any_block,
+                has_spec_block,
+            }) => {
+                let has_any_block = has_any_block || matches!(expr, Block(..) | Sequence(..));
+                let has_spec_block = has_spec_block || matches!(expr, SpecBlock(..));
+                self.state = Some(Traversing {
+                    id,
+                    has_any_block,
+                    has_spec_block,
+                });
+            },
+            _ => {},
+        }
+    }
+
+    fn visit_expr_post(&mut self, expr: &ExpData, env: &GlobalEnv) {
+        use CondExprState::*;
+        match self.state {
+            Some(Traversing {
+                id,
+                has_any_block,
+                has_spec_block,
+            }) if expr.node_id() == id => {
+                // We are done with traversing the condition of interest.
+                self.state = None;
+                if has_any_block && !has_spec_block {
+                    env.diag(
```

Comment:
We probably should add a new `env.lint_diag` function which gets passed the QualifiedId<FunId> of the function being processed, plus a name for the lint rule. The names of lint rules maybe `const &str` declared somewhere here in this directory. The `lint_diag` function would check whether rule is enabled/disabled. It would look at the attribute at the function, and then at the module. There is already a mechanism similar as this for prover pragmas (which were introduced before Move got attributes): https://github.com/aptos-labs/aptos-core/blob/0aba089be647caed11ff99e3aa6a87809f6c05d3/third_party/move/move-model/src/model.rs#L3697-L3706

Reply:
Good point, I was thinking of a similar mechanism. Please note that as mentioned in the PR description, this PR does not deal with the attribute-based silencing, but that feature will be added in a followup PR.

Reply:
If we land this now without this mechanism, do we get warnings in the frameworks? That wouldn't be good. If so, we need to lock any releases of the CLI until this is fixed.

Reply:
No, not unless you explicitly set the MVC_EXP variable. Lint checks are currently disabled by default and only exposed via env vars (as mentioned in the PR description).